DigitalLearningSolutions.Data.Tests/Services/GroupServiceTests/GroupsServiceTests.cs

```diff
@@ -662,6 +663,82 @@ public void
+                        centreId: reusableDelegateDetails.CentreId
+                    )
+                );
+
+            A.CallTo(
+                () => centreRegistrationPromptsService.GetCentreRegistrationPromptNameAndNumber(
+                    reusableDelegateDetails.CentreId,
+                    1
+                )
+            ).Returns("clique");
+            A.CallTo(
+                () => centreRegistrationPromptsService.GetCentreRegistrationPromptNameAndNumber(
+                    reusableDelegateDetails.CentreId,
+                    2
+                )
+            ).Returns("astronomy");
+            A.CallTo(
+                () => centreRegistrationPromptsService.GetCentreRegistrationPromptNameAndNumber(
+                    reusableDelegateDetails.CentreId,
+                    3
+                )
+            ).Returns("mathematics");
+
+            var group_to_join_1 = GroupTestHelper.GetDefaultGroup(48, "cool kids", linkedToField: 1);
+            var group_to_join_2 = GroupTestHelper.GetDefaultGroup(49, "astronomy - local group", linkedToField: 2);
+            var group_to_not_join = GroupTestHelper.GetDefaultGroup(49, "set with binary operation", linkedToField: 3);
+            var ineligibile_group = GroupTestHelper.GetDefaultGroup(
+                50,
+                "Bad - Apple",
+                shouldAddNewRegistrantsToGroup: false
+            );
+            var groups = new List<Group> { group_to_join_1, group_to_join_2, group_to_not_join, ineligibile_group };
+
+            var registrationFieldAnswers = new RegistrationFieldAnswers(
+                reusableDelegateDetails.CentreId,
+                reusableDelegateDetails.JobGroupId,
+                "cool kids",
+                "local group",
+                "noise",
+                null,
+                null,
+                null
+            );
+            A.CallTo(() => groupsDataService.GetGroupsForCentre(reusableDelegateDetails.CentreId)).Returns(groups);
+
+            A.CallTo(() => jobGroupsDataService.GetJobGroupName(reusableDelegateDetails.JobGroupId))
+                .Returns("job group name");
+
+            // when
+            groupsService.AddNewDelegateToRegistrationFieldGroupsAndEnrolOnCourses(
+                reusableDelegateDetails.Id,
+                registrationFieldAnswers
+            );
+
+            // then
+            A.CallTo(
+                () => groupsDataService.AddDelegateToGroup(reusableDelegateDetails.Id, group_to_join_1.GroupId, now, 1)
+            ).MustHaveHappenedOnceExactly();
+            A.CallTo(
+                () => groupsDataService.AddDelegateToGroup(reusableDelegateDetails.Id, group_to_join_2.GroupId, now, 1)
+            ).MustHaveHappenedOnceExactly();
+            A.CallTo(
+                () => groupsDataService.AddDelegateToGroup(A<int>._, A<int>._, A<DateTime>._, A<int>._)
+            ).MustHaveHappenedTwiceExactly();
+        }
+

```

Comment:
Add more tests:
* That `AddNewDelegateToRegistrationFieldGroupsAndEnrolOnCourses` calls `progressDataService.CreateNewDelegateProgress` and `progressDataService.CreateNewAspProgress` when appropriate
* That `AddNewDelegateToRegistrationFieldGroupsAndEnrolOnCourses` calls `progressDataService.UpdateProgressSupervisorAndCompleteByDate` when appropriate
* That `AddNewDelegateToRegistrationFieldGroupsAndEnrolOnCourses` schedules an enrolment email

@SteveJacksonSoft I see that in GroupService, we do [either UpdateProgressSupervisorAndCompleteByDate or CreateNewDelegateProgress](https://github.com/TechnologyEnhancedLearning/DLSV2/blob/55f5fb65038bbf7335dc5c0874a3d6745e392fbb/DigitalLearningSolutions.Data/Services/GroupsService.cs#L610) but never both. Is this correct?